framework/src/fviks/FVScalarLagrangeMultiplierInterface.C

```diff
@@ -0,0 +1,123 @@
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "FVScalarLagrangeMultiplierInterface.h"
+#include "MooseVariableScalar.h"
+
+registerMooseObject("MooseApp", FVScalarLagrangeMultiplierInterface);
+
+InputParameters
+FVScalarLagrangeMultiplierInterface::validParams()
+{
+  InputParameters params = FVInterfaceKernel::validParams();
+  params.addClassDescription(
+      "This class should be inherited to create interface penalties in finite volume.");
+  params.addRequiredCoupledVar("lambda", "The name of the scalar lagrange multiplier");
+
+  return params;
+}
+
+FVScalarLagrangeMultiplierInterface::FVScalarLagrangeMultiplierInterface(
+    const InputParameters & params)
+  : FVInterfaceKernel(params),
+    _lambda_var(*getScalarVar("lambda", 0)),
+    _lambda(adCoupledScalarValue("lambda"))
+{
+}
+
+ADReal
+FVScalarLagrangeMultiplierInterface::computeQpResidual()
+{
+  mooseError("There is no need for me.");
+}
+
+void
+FVScalarLagrangeMultiplierInterface::computeResidual(const FaceInfo & fi)
+{
+  _face_info = &fi;
+  _normal = fi.normal();
+  _elem_is_one = _subdomain1.find(fi.elem().subdomain_id()) != _subdomain1.end();
+
+#ifndef NDEBUG
+  const auto ft1 = fi.faceType(_var1.name());
+  const auto ft2 = fi.faceType(_var2.name());
+  constexpr auto ft_both = FaceInfo::VarFaceNeighbors::BOTH;
+  constexpr auto ft_elem = FaceInfo::VarFaceNeighbors::ELEM;
+  constexpr auto ft_neigh = FaceInfo::VarFaceNeighbors::NEIGHBOR;
+  mooseAssert(((ft1 == ft_both) && (ft2 == ft_both)) ||
+                  (_elem_is_one && (ft1 == ft_elem) && (ft2 == ft_neigh)) ||
+                  (!_elem_is_one && (ft1 == ft_neigh) && (ft2 == ft_elem)),
+              "These seem like the reasonable combinations of face types.");
+#endif
+  mooseAssert(_local_re.size() == 1, "We should only have a single dof");
+  mooseAssert(_lambda.size() == 1 && _lambda_var.order() == 1,
+              "The lambda variable should be first order");
+  mooseAssert(_lambda_var.dofIndices().size() == 1, "We should only have a single dof");
+
+  const auto var_elem_num = _elem_is_one ? _var1.number() : _var2.number();
+  const auto var_neigh_num = _elem_is_one ? _var2.number() : _var1.number();
+
+  const auto r = MetaPhysicL::raw_value(_lambda[0]) * fi.faceArea() * fi.faceCoord();
+
+  // Primal residual
+  prepareVectorTag(_assembly, var_elem_num);
+  _local_re(0) = r;
+  accumulateTaggedLocalResidual();
+  prepareVectorTagNeighbor(_assembly, var_neigh_num);
+  _local_re(0) = -r;
+  accumulateTaggedLocalResidual();
+
+  // LM residual. We may not have any actual ScalarKernels in our simulation so we need to manually
+  // make sure the scalar residuals get cached for later addition
+  const auto lm_r = MetaPhysicL::raw_value(computeQpResidual()) * fi.faceArea() * fi.faceCoord();
```

Comment:
I'm guessing you have a more complex problem than this... one problem I see is that right now this would raise a `mooseError` :laughing:

Reply:
lol I ll delete that definition. It's virtual though it's overwritten in the interface kernel I care about

Reply:
ah it wasnt even virtual. Well now it's pure virtual so we are safe.